src/arm/windows/init_by_logical_sys_info.c

```diff
@@ -0,0 +1,842 @@
+		break;
+		case 2:
+			current_cache = l2_base + l2_counter;
+			l2_counter++;
+		break;
+		case 3:
+			current_cache = l3_base + l3_counter;
+			l3_counter++;
+		break;
+	}
+	current_cache->size = info->Cache.CacheSize;
+	current_cache->line_size = info->Cache.LineSize;
+	current_cache->associativity = info->Cache.Associativity;
+	/* We don't have partition and set information of caches on Windows,
+	 * so we set partitions to 1 and calculate the expected sets.
+	 */
+	current_cache->partitions = 1;
+	current_cache->sets =
+		current_cache->size / current_cache->line_size / current_cache->associativity;
+	if (info->Cache.Type == CacheUnified) {
+		current_cache->flags = CPUINFO_CACHE_UNIFIED;
+	}
+
+	for (uint32_t i = 0; i <= info->Cache.GroupCount; i++) {
+	/* Zero GroupCount is valid, GroupMask still can store bits set. */
+		const uint32_t group_id = info->Cache.GroupMasks[i].Group;
+		/* Bitmask representing processors in this group belonging to this package */
+		KAFFINITY group_processors_mask = info->Cache.GroupMasks[i].Mask;
+		while (group_processors_mask != 0) {
+			const uint32_t processor_id_in_group =
+				low_index_from_kaffinity(group_processors_mask);
+			const uint32_t processor_global_index =
+				global_proc_index_per_group[group_id] + processor_id_in_group;
+
+			store_cache_info_per_processor(
+				processors, processor_global_index,
+				info, current_cache);
+
+			/* Reset the lowest bit in affinity mask */
+			group_processors_mask &= (group_processors_mask - 1);
+		}
+	}
+}
+
+static void store_package_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	const uint32_t package_id,
+	const uint32_t group_id,
+	const uint32_t processor_id_in_group,
+	const struct woa_chip_info *chip_info)
+{
+	processors[processor_global_index].windows_group_id =
+		(uint16_t) group_id;
+	processors[processor_global_index].windows_processor_id =
+		(uint16_t) processor_id_in_group;
+
+	/* As we're counting the number of packages now, we haven't allocated memory for */
+	/* cpuinfo_packages yet, so we only set the package pointer's offset now. */
+	processors[processor_global_index].package =
+		(const struct cpuinfo_package*) NULL + package_id;
+}
+
+void store_core_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	const uint32_t core_id,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX core_info,
+	struct cpuinfo_core* cores,
+	const struct woa_chip_info *chip_info)
+{
+	if (cores) {
+		processors[processor_global_index].core = cores + core_id;
+		cores[core_id].core_id = core_id;
+		get_core_uarch_for_efficiency(
+			chip_info->chip_name, core_info->Processor.EfficiencyClass,
+			&(cores[core_id].uarch), &(cores[core_id].frequency));
+
+		/* We don't have cluster information, so we handle it as
+		 * fixed 1 to (cluster / cores).
+		 */
+		processors[processor_global_index].cluster =
+			(const struct cpuinfo_cluster*) NULL + core_id;

```

Comment:
I dont understand what this is doing or how its safe. cluster as a pointer is being set to an absolute address equivalent to the core_id * sizeof(cpuinfo_cluster). Did I miss something?

Reply:
I figured it out from the code later that patches these values by adding a base offset. Can you add a comment here to make it clear?

Reply:
Again, I've copied this pattern from x86: https://github.com/pytorch/cpuinfo/blob/master/src/x86/windows/init.c#L224
I've struggled a lot to understand why it's useful and I've tried to add as many as clear as I could, but as you've struggled also, I've failed with that!